crates/noirc_evaluator/src/lib.rs

```diff
@@ -91,7 +102,8 @@ impl Evaluator {
         // an intermediate variable.
         let is_intermediate_variable = witness_index.as_usize() > self.num_witnesses_abi_len;
 
-        let is_public_input = self.public_inputs.contains(&witness_index);
+        let is_public_input =
```

Comment:
Are all `param_witnesses` public inputs?

Reply:
ie what If I did :

```rust
fn main(x : Field, y : pub Field) {}
```

Where would `x` be stored?

Reply:
Ah, good spot. `param_witnesses` holds both public and private inputs so this is a bug.

Reply:
I noticed you are no longer passing `AbiVisibility` to `param_to_var` below. Could change this back and then have the `param_witnesses` map be `BTreeMap<String, (AbiVisibility, Vec<Witness>)>`, or bring back `public_inputs` in the evaluator just for this method

Reply:
I've re-added the `public_inputs` field as this simplifies passing this value into the `Circuit` as well.

We should be able to fairly easily replace this with a `BTreeSet` which would remove all of the deduping we have to do up in nargo. We'll also be able to strip out all of the array variants of ABI encoding.

I'm not sure this PR is the place for this however (don't want to cause churn on already reviewed code) but I can follow up with a simplification PR.

Reply:
I think a follow up PR is better

Reply:
> I noticed you are no longer passing `AbiVisibility` to `param_to_var` below. Could change this back and then have the `param_witnesses` map be `BTreeMap<String, (AbiVisibility, Vec<Witness>)>`, or bring back `public_inputs` in the evaluator just for this method

@vezenovm why did you need this btw?

Reply:
He was just suggesting ways to address this comment.